File: rails/app/models/deck.rb
class Deck < ApplicationRecord
  self.table_name = 'decks'

  enum :format, { standard: 0, extended: 1, legacy: 2, vintage: 3, commander: 4, draft: 5 }, prefix: :format
  enum :archetype, { aggro: 0, control: 1, midrange: 2, combo: 3, prison: 4, tempo: 5 }, prefix: :archetype

  belongs_to :player, class_name: 'Player'
  has_many :cards, class_name: 'Card', through: :deck_cards
  has_many :sideboard_cards, class_name: 'Card', through: :deck_sideboard_cards
  has_many :tags, class_name: 'DeckTag', through: :deck_tag_assignments

  validates :name, presence: true, length: { maximum: 100 }

  # Domain invariants — simple rules
  validate :validate_rules

  def validate_rules
    errors.add(:wins_not_negative, 'Deck wins count must not be negative') unless ((wins.nil? || wins >= 0))
    errors.add(:losses_not_negative, 'Deck losses count must not be negative') unless ((losses.nil? || losses >= 0))
    errors.add(:draws_not_negative, 'Deck draws count must not be negative') unless ((draws.nil? || draws >= 0))
  end

  # Domain invariants — IMPLIES rules
  validate :validate_implies

  def validate_implies
    errors.add(:base, 'Tournament-legal deck must be made public') if (is_tournament_legal == true) && !(is_public == true)
  end

  def to_s
    name.to_s
  end

  # Business operations

  def validate_size
    # TODO: implement validate_size
    nil
  end

  def add_card(card_id, quantity)
    # TODO: implement add_card
  end

  def remove_card(card_id)
    # TODO: implement remove_card
  end

  def win_rate
    # TODO: implement win_rate
    nil
  end

  def clone
    # TODO: implement clone
    nil
  end

  def publish
    # TODO: implement publish
  end

  def unpublish
    # TODO: implement unpublish
  end

  def certify_tournament_legal
    # TODO: implement certify_tournament_legal
    nil
  end

  # Lifecycle hooks
  after_save :recalculate_tournament_legal

  def recalculate_tournament_legal
    # TODO: implement recalculate_tournament_legal
  end

  def as_json(options = {})
    hash = super(options)
    hash['createdAt'] = hash.delete('created_at') if hash.key?('created_at')
    hash['updatedAt'] = hash.delete('updated_at') if hash.key?('updated_at')
    hash
  end
end

File: dotnet/CardsProject/Domain/Cards/Deck.cs
using CardsProject.Domain.Players;
using System.ComponentModel.DataAnnotations.Schema;
using System.ComponentModel.DataAnnotations;
using System.Text.Json.Serialization;

namespace CardsProject.Domain.Cards;

public enum DeckFormatType
{
    Standard,
    Extended,
    Legacy,
    Vintage,
    Commander,
    Draft
}

public enum DeckArchetypeType
{
    Aggro,
    Control,
    Midrange,
    Combo,
    Prison,
    Tempo
}

public class Deck : IValidatableObject
{
    public int Id { get; set; }

    public string Name { get; set; } = "";
    public string? Description { get; set; }
    public DeckFormatType Format { get; set; }
    public bool IsPublic { get; set; } = false;
    public bool IsTournamentLegal { get; set; } = false;
    public DeckArchetypeType? Archetype { get; set; }
    public int Wins { get; set; } = 0;
    public int Losses { get; set; } = 0;
    public int Draws { get; set; } = 0;
    [JsonPropertyName("createdAt")]
    public DateTime? CreatedAt { get; set; } = null;
    [JsonPropertyName("updatedAt")]
    public DateTime? UpdatedAt { get; set; } = null;

    public int? PlayerId { get; set; }
    [ForeignKey(nameof(PlayerId))]
    public Player? Player { get; set; }

    public ICollection<Card> Cards { get; set; } = new List<Card>();
    public ICollection<Card> SideboardCards { get; set; } = new List<Card>();
    public ICollection<DeckTag> Tags { get; set; } = new List<DeckTag>();

    // Business operations

    public bool ValidateSize()
    {
        // TODO: implement validate_size
        return default;
    }

    public void AddCard(int cardId, int quantity)
    {
        // TODO: implement add_card
    }

    public void RemoveCard(int cardId)
    {
        // TODO: implement remove_card
    }

    public decimal WinRate()
    {
        // TODO: implement win_rate
        return default;
    }

    public object Clone()
    {
        // TODO: implement clone
        return default;
    }

    public void Publish()
    {
        // TODO: implement publish
    }

    public void Unpublish()
    {
        // TODO: implement unpublish
    }

    public bool CertifyTournamentLegal()
    {
        // TODO: implement certify_tournament_legal
        return default;
    }

    // ── Domain invariants (simple rules) ──────────────────────────────
    public IEnumerable<ValidationResult> Validate(ValidationContext validationContext)
    {
        if (!( Wins >= 0 ))
            yield return new ValidationResult("Deck wins count must not be negative", new[] { nameof(Id) });
        if (!( Losses >= 0 ))
            yield return new ValidationResult("Deck losses count must not be negative", new[] { nameof(Id) });
        if (!( Draws >= 0 ))
            yield return new ValidationResult("Deck draws count must not be negative", new[] { nameof(Id) });
    }

    // ── Lifecycle hooks (call from AppDbContext.SaveChangesAsync override) ───
    public void RecalculateTournamentLegal()
    {
        // TODO: implement recalculate_tournament_legal
    }
}
